More comprehensive YAML config errors + dynamic checks for action configs

spacegaier opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Breaking change

<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

## Proposed change

To help the users (and prevent questions in Discord 😄), this PR improves the config checks, both in general as well as specifically for action configs.

**General:**
1. Checks for missing keys (shown in red, same as technical YAML structure errors).
2. Multiple errors can now be shown (e.g. if multiple fields are missing).
3. Bump `superstruct` version since only from 0.10.13 on does the library provide multiple `failures` so we have multiple error messages.
4. Incorporates the spelling changes from PR https://github.com/home-assistant/frontend/pull/8211 (but now properly translated) + replaced hardcoded strings with translations.

**Action configs:**
1. Make use of `dynamic` checks in `superstruct` so we can dynamically enforce the properties depending on the selected action type.

![image](https://user-images.githubusercontent.com/114137/105618533-a1022400-5de8-11eb-97b3-b49fdf9c834d.png)

![action_validation](https://user-images.githubusercontent.com/114137/105628019-8bb5e580-5e3a-11eb-9e17-e37f95b6ea8a.gif)



## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

spacegaier committed:
Implement stricter YAML config checks

spacegaier committed:
Remove dummy property

spacegaier committed:
Safeguards + styling tweak

spacegaier committed:
Proper dynamic setup + confirmation handling

spacegaier committed:
Rendering and confirmation fixes

spacegaier committed:
Added translations

bramkragten commented on src/panels/config/automation/action/ha-automation-action-row.ts:
```diff
@@ -251,7 +251,7 @@ export default class HaAutomationActionRow extends LitElement {
 
   private _handleUiModeNotAvailable(ev: CustomEvent) {
-    this._warnings = handleStructError(ev.detail);
+    this._warnings = handleStructError(this.hass, ev.detail).warnings;
```
The automation editor is using helpers of Lovelace, that is not correct, if it is a global helper move it out of Lovelace

spacegaier committed:
Properly handle missing or unsupported types

spacegaier commented:
@bramkragten Do you want to check again? I moved the sources files as requested and also adjusted the logic a bit, since handling of unsupported types was wonky (also before my changes). We never properly detected it, once you made changes to he YAML.

spacegaier committed:
Revert unintended change

bramkragten commented on src/panels/lovelace/editor/hui-element-editor.ts:
```diff
@@ -66,11 +66,13 @@ export abstract class HuiElementEditor<T> extends LitElement {
   // Warning: GUI editor can't handle configuration - ok to save
   @internalProperty() private _warnings?: string[];
 
+  @internalProperty() private _typeSupported?: boolean;
```
```suggestion
  @internalProperty() private _guiSupported?: boolean;
```

bramkragten commented on src/panels/lovelace/editor/hui-element-editor.ts:
```diff
@@ -194,29 +203,44 @@ export abstract class HuiElementEditor<T> extends LitElement {
               </div>
             `}
-        ${this._error
+        ${!this._typeSupported && this.configElementType
```
```suggestion
        ${this._guiSupported === false
```

spacegaier replied:
Without the check for `configElementType` we would also show the warning if no type is provided. But for that we already have a red error message.

bramkragten replied:
But now we also show it before the config element is set? Because `_guiSupported` would be undefined?

spacegaier replied:
Now I got your point. Adjusted plus fixed an issue with the "Save" button.

spacegaier committed:
Changes from review

spacegaier committed:
Changes from review
